src/objective/regression_obj.cu

```diff
@@ -73,29 +73,43 @@ class RegLossObj : public ObjFunction {
     additional_input_.HostVector().begin()[1] = scale_pos_weight;
     additional_input_.HostVector().begin()[2] = is_null_weight;
 
-    common::Transform<>::Init([] XGBOOST_DEVICE(size_t _idx,
+    const size_t nthreads = omp_get_max_threads();
+    const size_t n_data_blocks = nthreads;
```

Comment:
Can we simply use the raw pointer optimization and remove the thread block?  This function is universal to CPU and GPU. (please don't make a separated implementation ;-) )

Reply:
Or alternatively you can modify the transform function.  But I prefer to keep things simple considering the objective should be a small fraction of overall runtime.

Reply:
seems no need to modify transform func, added simple support of old behaviour for gpu device (`n_data_blocks == ndata`)